src/aliasthis.c

```diff
@@ -21,18 +21,376 @@
+        if (d && !d->isTupleDeclaration())
+        {
+            assert(d->type);
+            Type *t2 = d->type;
+            if (d->isVarDeclaration() && d->needThis())
+            {
+                t2 = t2->addMod(t->mod);
+                *islvalue = true; //Variable is always l-value
+            }
+            else if (d->isFuncDeclaration())
+            {
+                FuncDeclaration *fd = resolveFuncCall(Loc(), NULL, d, NULL, t, NULL, 1);
+                if (fd && fd->errors)
+                    return Type::terror;
+                if (fd && !fd->type->nextOf() && !fd->functionSemantic())
+                    fd = NULL;
+                if (fd)
+                {
+                    t2 = fd->type->nextOf();
+                    if (!t2) // issue 14185
+                        return Type::terror;
+                    t2 = t2->substWildTo(t->mod == 0 ? MODmutable : t->mod);
+                    if (((TypeFunction *)fd->type)->isref)
+                        *islvalue = true;
+                }
+                else
+                    return Type::terror;
+            }
+            return t2;
+        }
+        EnumDeclaration *ed = s->isEnumDeclaration();
+        if (ed)
+        {
+            Type *t2 = ed->type;
+            return t2;
+        }
+        TemplateDeclaration *td = s->isTemplateDeclaration();
+        if (td)
+        {
+            assert(td->scope);
+            FuncDeclaration *fd = resolveFuncCall(Loc(), NULL, td, NULL, t, NULL, 1);
+            if (fd && fd->errors)
+                return Type::terror;
+            if (fd && fd->functionSemantic())
+            {
+                Type *t2 = fd->type->nextOf();
+                t2 = t2->substWildTo(t->mod == 0 ? MODmutable : t->mod);
+                if (((TypeFunction *)fd->type)->isref)
+                    *islvalue = true;
+                return t2;
+            }
+            else
+                return Type::terror;
+        }
+        //printf("%s\n", s->kind());
+    }
+    return NULL;
+}
+
+/**
+    This function walk over `from` basetype tree and search types,
+    which can be converted (without alias this subtyping) to `to`.
+    If there are many ways to convert `from` to `to`, this function
+    raises an error and prints all those ways.
+    To prevent infinity loop in types with circular subtyping,
+    pattern "check a flag, lock a flag, do work is a flag wasn't locked, return flag back" is used.
+*/
+MATCH implicitConvToWithAliasThis(Loc loc, Type *from, Type *to, Type *root_from, OutBuffer *buff,
+                                  int *state, OutBuffer *matchname)
+{
+    //printf("implicitConvToWithAliasThis, %s -> %s\n", from->toChars(), to->toChars());
+    if (from->aliasthislock & RECtracing)
+        return MATCHnomatch;
+
+    unsigned oldatlock = from->aliasthislock;
+    from->aliasthislock |= RECtracing;
+    if (!buff)
+    {
+        buff = new OutBuffer();
+        buff->writestring(from->toChars());
```

Comment:
This buff is needed to the properly error messages: 
We need to know name of the currect alias this symbol (`Foo.bar.boo` or `cast(ParentBar)(Foo.bar).moo`.)
And If we have found the one match, we should store the matching symbol name and try to find another matches. 
I'll reflect it in the comment.